apps/analysis/log.py

```diff
@@ -0,0 +1,82 @@
+import dataclasses
```

Comment:
This looks a lot like python's logger. Was it easier / quiker to implement this than to try to piggy-back on the logger's functionality?

Reply:
Originally this was going more but can probably remove it now.